pkg/test/kube/forwarder.go

```diff
@@ -1,107 +1,184 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
 
 package kube
 
 import (
-	"context"
+	"bytes"
+	"errors"
 	"fmt"
-	"os/exec"
+	"net/http"
+	"os"
 	"strings"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	kubelib "istio.io/istio/pkg/kube"
 )
 
-// TODO: Consider replacing this code with a direct call to the underlying client library, or wrapping
-// the executable call in a shell function, so that the process doesn't end up running indefinitely in case
-// of abrupt end to the test run.
-// See https://github.com/istio/istio/issues/6173
-
-// PortForwarder represents a locally started kubectl process that forwards a local port to one within the
-// cluster.
-type PortForwarder struct {
-	ctx     context.Context
-	cancel  context.CancelFunc
-	cmd     *exec.Cmd
-	address string
-	target  string
+// PortForwarder is a wrapper for port forward.
+type PortForwarder interface {
+	ForwardPorts() error
+	Address() string
+	Close()
 }
 
-// NewPortForwarder returns a new PortForwarder the specified port in the given namespace/pod.
-func NewPortForwarder(kubeconfig string, ns string, pod string, port int) *PortForwarder {
-	command := fmt.Sprintf("kubectl port-forward --kubeconfig=%s -n %s %s :%d", kubeconfig, ns, pod, port)
+type defaultPortForwarder struct {
+	forwarder *portforward.PortForwarder
+	readyCh   <-chan struct{}
+	address   string
+	output    *bytes.Buffer
+}
+
+// PodSelectOptions contains options for pod selection.
+// Must specify PodNamespace and one of PodName/LabelSelector.
+type PodSelectOptions struct {
+	PodNamespace  string
+	PodName       string
+	LabelSelector string
+}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	cmd := exec.CommandContext(ctx, "sh", "-c", command)
+func (f *defaultPortForwarder) ForwardPorts() error {
+	errCh := make(chan error)
+	go func() {
+		errCh <- f.forwarder.ForwardPorts()
+	}()
 
-	return &PortForwarder{
-		ctx:    ctx,
-		cancel: cancel,
-		cmd:    cmd,
-		target: fmt.Sprintf("%s/%s (%d)", ns, pod, port),
+	select {
+	case err := <-errCh:
+		return fmt.Errorf("failure running port forward process: %v", err)
+	case <-f.readyCh:
+		address, err := extractAddress(f.output.String())
+		if err != nil {
+			return err
+		}
+		f.address = address
+		return nil
 	}
 }
 
-// Address is the local address for the port forwarder.
-func (p *PortForwarder) Address() string {
-	return p.address
+func (f *defaultPortForwarder) Address() string {
+	return f.address
+}
+
+func (f *defaultPortForwarder) Close() {
+	f.forwarder.Close()
```

Comment:
Does this send something to the `stopCh`? If so, maybe we keep this internal and force the caller to use `Close`?

Reply:
Not really. the Close func here will close all of the opening listeners for portforward,and then lead to the forward go routine to return.
And if the stopCh is closed/triggered, the ForwardPorts() process will return, and entire streamConnection will be destroyed.
There will be no orphan portforward process for both of the two scenarioes. The stopCh is mainly for the scenario of unexpected panic or terminate

Reply:
Oh, forget that, I made the thing complex.
So, yeah, we don't need to expose the `stopCh` to end users